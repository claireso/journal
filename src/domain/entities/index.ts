export * from './photo'
export * from './subscription'
export * from './media'
export * from './user'
export * from './pager'
