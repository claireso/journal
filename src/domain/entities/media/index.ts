export * from './media'
export * from './request'
export * from './helpers'
