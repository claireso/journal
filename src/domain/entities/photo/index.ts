export * from './photo'
export * from './photos'
export * from './mappers'
