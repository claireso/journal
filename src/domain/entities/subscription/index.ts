export * from './subscription'
export * from './subscriptions'
