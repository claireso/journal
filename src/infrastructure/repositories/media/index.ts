export { default as MediaRepositoryImpl } from './MediaRepositoryImpl'
