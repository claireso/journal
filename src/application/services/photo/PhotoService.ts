import escape from 'lodash/escape'
import { PhotoRepository, MediaRepository } from '@domain/repositories'
import { Pager, Photos } from '@domain/entities'
import { BadRequestError, NotFoundError } from '@domain/errors/errors'
import { PhotoInsertDto, PhotoUpdateDto } from '@dto'

export default class PhotoService {
  private repository: PhotoRepository
  private mediaRepository: MediaRepository
  private logger: any

  constructor(repository: PhotoRepository, mediaRepository: MediaRepository, logger: any) {
    this.repository = repository
    this.mediaRepository = mediaRepository
    this.logger = logger
  }

  private _cleanPhotoData(data: any) {
    return {
      title: escape(data.title),
      description: escape(data.description),
      color: data.color || null,
      position: data.position,
      media_id: data.media_id,
      updated_at: new Date()
    }
  }

  async create(data: PhotoInsertDto) {
    // check if the media exists
    const media = await this.mediaRepository.getById(data.media_id)
    if (media === null) {
      throw new BadRequestError(`Can not create photo because media does not exist`, {
        cause: { data }
      })
    }

    // check if the media isn't linked to a photo
    const linkedPhoto = await this.repository.getByMediaId(media.id)
    if (linkedPhoto !== null) {
      throw new BadRequestError(`Can not create photo because media is already linked to a photo`, {
        cause: { data }
      })
    }

    const cleanedData = this._cleanPhotoData(data)

    return this.repository.create({
      ...cleanedData,
      name: media.name // set a value for "name" to keep legacy working (a database migration is required in the future)
    })
  }

  async update(id: number, data: PhotoUpdateDto) {
    const photo = await this.getById(id)

    const newPhoto = this._cleanPhotoData({
      title: data.title ?? photo.title,
      description: data.description ?? photo.description,
      media_id: data.media_id || photo.media_id,
      color: data.color ?? photo.color,
      position: data.position ?? photo.position
    })

    // The media will be updated
    if (newPhoto.media_id !== photo.media_id) {
      // check if the media exists
      const media = await this.mediaRepository.getById(newPhoto.media_id)
      if (media === null) {
        throw new BadRequestError(`Can not update photo because media does not exist`, {
          cause: { data }
        })
      }

      // check if the media isn't linked to a photo
      const linkedPhoto = await this.repository.getByMediaId(media.id)
      if (linkedPhoto !== null) {
        throw new BadRequestError(`Can not create photo because media is already linked to a photo`, {
          cause: { data }
        })
      }
    }

    return this.repository.update(id, newPhoto)
  }

  async getById(id: number) {
    const photo = await this.repository.getById(id)

    if (photo === null) {
      throw new NotFoundError(`Not found photo`, { cause: { photoId: id } })
    }

    return photo
  }

  async getByMediaId(mediaId: number) {
    return this.repository.getByMediaId(mediaId)
  }

  async getPreviousPhoto() {
    return this.repository.getPreviousPhoto()
  }

  async delete(id: number, db: any) {
    const photo = await this.getById(id)

    // on delete photo, delete linked media
    try {
      // start transaction
      // todo: find a way to manage db transactions
      await db.query('BEGIN')
      // delete photo from database
      await this.repository.delete(id)

      if (photo.media_id) {
        // delete media from database
        // todo: delete media file
        await this.mediaRepository.delete(photo.media_id)
      }
      await db.query('COMMIT')
    } catch (e) {
      await db.query('ROLLBACK')
      throw e
    }
  }

  async getPaginatedPhotos(page: number) {
    const pageSize = 10

    const count = await this.repository.countPhotos()
    const totalPages = Math.ceil(count / pageSize)

    if (page < 1 || (totalPages && page > totalPages)) {
      throw new NotFoundError('Page photo not found', { cause: { page } })
    }

    const offset = (page - 1) * pageSize

    const photos = await this.repository.getPhotos(offset, pageSize)

    const pager: Partial<Pager> = {
      count,
      totalPages,
      offset,
      limit: pageSize
    }

    if (page > 1) {
      pager.prev = page - 1
      pager.first = 1
    }

    if (page < totalPages) {
      pager.next = page + 1
      pager.last = totalPages
    }

    return {
      items: photos,
      pager
    } as Photos
  }
}
