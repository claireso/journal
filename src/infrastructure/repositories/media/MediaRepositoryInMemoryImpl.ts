import { Media, MediaType, mapRowToMedia } from '@domain/entities'
import { MediaRepository } from '@domain/repositories'

export default class MediaRepositoryInMemoryImpl implements MediaRepository {
  private media: Media[]

  constructor() {
    this.media = [
      {
        id: 1,
        type: MediaType.IMAGE,
        name: 'abc.jpg',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        size: {
          width: 1640,
          height: 1093
        }
      },
      {
        id: 2,
        type: MediaType.IMAGE,
        name: 'def.jpg',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        size: {
          width: 800,
          height: 1200
        }
      },
      {
        id: 3,
        type: MediaType.IMAGE,
        name: 'ghi.jpg',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        size: {
          width: 800,
          height: 800
        }
      },
      {
        id: 4,
        type: MediaType.IMAGE,
        name: 'jkl.jpg',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        size: {
          width: 1640,
          height: 1093
        }
      }
    ]
  }

  async create(data: Pick<Media, 'type' | 'name' | 'size'>): Promise<Media> {
    const {
      type,
      name,
      size: { width, height }
    } = data

    const lastMedia = this.media.at(-1) as Media
    const newMedia = mapRowToMedia({
      id: lastMedia.id + 1,
      created_at: new Date('2024-10-05T17:12:40.400Z'),
      type,
      name,
      width,
      height
    })
    this.media.push(newMedia)
    return newMedia
  }

  async getById(id: number): Promise<Media | null> {
    const media = this.media.find((_media) => _media.id === id)
    return media || null
  }

  async delete(id: number): Promise<void> {
    this.media = this.media.filter((_media) => _media.id !== id)
  }
}
