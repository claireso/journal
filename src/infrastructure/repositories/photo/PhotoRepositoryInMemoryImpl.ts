import { PhotoInsertDto, PhotoUpdateDto } from '@dto'
import { MediaType, Photo, PhotoPositionType } from '@domain/entities'
import { PhotoRepository } from '@domain/repositories'
import { mapRowToPhoto } from '@domain/entities'

export default class PhotoRepositoryInMemoryImpl implements PhotoRepository {
  private photos: Photo[]

  constructor() {
    this.photos = [
      {
        id: 1,
        name: 'abc.jpg',
        title: '',
        description: '',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        updated_at: new Date('2024-10-05T17:12:40.400Z'),
        position: PhotoPositionType.LEFT,
        media_id: 1,
        media: {
          id: 1,
          type: MediaType.IMAGE,
          name: 'abc.jpg',
          created_at: new Date('2024-10-05T17:12:40.400Z'),
          size: {
            width: 1640,
            height: 1093
          }
        },
        color: null,
        portrait: false,
        square: false
      },
      {
        id: 2,
        name: 'def.jpg',
        title: '',
        description: '',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        updated_at: new Date('2024-10-05T17:12:40.400Z'),
        position: PhotoPositionType.CENTER,
        media_id: 2,
        media: {
          id: 2,
          type: MediaType.IMAGE,
          name: 'def.jpg',
          created_at: new Date('2024-10-05T17:12:40.400Z'),
          size: {
            width: 800,
            height: 1200
          }
        },
        color: null,
        portrait: true,
        square: false
      },
      {
        id: 3,
        name: 'ghi.jpg',
        title: '',
        description: '',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        updated_at: new Date('2024-10-05T17:12:40.400Z'),
        position: PhotoPositionType.CENTER,
        media_id: 3,
        media: {
          id: 3,
          type: MediaType.IMAGE,
          name: 'ghi.jpg',
          created_at: new Date('2024-10-05T17:12:40.400Z'),
          size: {
            width: 800,
            height: 800
          }
        },
        color: null,
        portrait: false,
        square: true
      }
    ]
  }

  async create(data: PhotoInsertDto & { name: string }): Promise<Photo> {
    const lastPhoto = this.photos.at(-1) as Photo

    const newPhoto = mapRowToPhoto({
      photo_id: lastPhoto.id + 1,
      title: data.title,
      description: data.description,
      position: data.position,
      color: data.color,
      photo_created_at: new Date('2024-10-05T17:12:40.400Z'),
      updated_at: new Date('2024-10-05T17:12:40.400Z'),
      media_id: 4,
      media_created_at: new Date('2024-10-05T17:12:40.400Z'),
      media_type: 'image',
      media_name: data.name,
      media_width: 1640,
      media_height: 1093,
      // deprecated fields
      photo_name: data.name,
      portrait: false,
      square: false
    })

    this.photos.push(newPhoto)
    return newPhoto
  }

  async update(id: number, data: PhotoUpdateDto): Promise<Photo> {
    this.photos = this.photos.map((photo) => {
      if (photo.id !== id) {
        return { ...photo }
      }

      return {
        ...photo,
        ...data
      }
    })

    return this.photos.find((photo) => photo.id === id) as Photo
  }

  async getById(id: number): Promise<Photo | null> {
    const photo = this.photos.find((photo) => photo.id === id)
    return photo || null
  }

  async getByMediaId(mediaId: number): Promise<Photo | null> {
    const photo = this.photos.find((photo) => photo.media_id === mediaId)
    return photo || null
  }

  async getPreviousPhoto(): Promise<Photo | null> {
    const photo = this.photos.at(-1)
    return photo || null
  }

  async delete(id: number): Promise<void> {
    this.photos = this.photos.filter((photo) => photo.id !== id)
  }

  async getPhotos(offset: number, limit: number) {
    const photos = this.photos.slice(offset, limit)
    return photos
  }

  async countPhotos(): Promise<number> {
    return this.photos.length
  }
}
