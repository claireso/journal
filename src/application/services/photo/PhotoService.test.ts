import { PhotoPositionType } from '@domain/entities'
import { MediaRepository, PhotoRepository } from '@domain/repositories'
import PhotoRepositoryInMemoryImpl from '@infrastructure/repositories/photo/PhotoRepositoryInMemoryImpl'
import MediaRepositoryInMemoryImpl from '@infrastructure/repositories/media/MediaRepositoryInMemoryImpl'
import PhotoService from './PhotoService'

describe('application/PhotoService', () => {
  let photoRepository: PhotoRepository
  let mediaRepository: MediaRepository
  let photoService: PhotoService

  beforeEach(() => {
    mediaRepository = new MediaRepositoryInMemoryImpl()
    photoRepository = new PhotoRepositoryInMemoryImpl()
    photoService = new PhotoService(photoRepository, mediaRepository, console)
  })

  describe('CREATE', () => {
    it('should create a photo successfully', async () => {
      // arrange
      const data = {
        title: 'new photo',
        description: 'new photo description',
        position: PhotoPositionType.LEFT,
        color: null,
        media_id: 4
      }

      // act
      const photo = await photoService.create(data)

      // assert
      expect(photo).toMatchSnapshot()
    })

    it('should throw a BadRequestError when creating a media with invalid media_id (media not found)', async () => {
      // arrange
      const data = {
        title: 'new photo',
        description: 'new photo description',
        position: PhotoPositionType.LEFT,
        color: null,
        media_id: 1000
      }

      // act
      const promise = photoService.create(data)

      // assert
      expect(promise).rejects.toThrow('Can not create photo because media does not exist')
    })

    it('should throw a BadRequestError when creating a media with invalid media_id (media already linked with a photo)', async () => {
      // arrange
      const data = {
        title: 'new photo',
        description: 'new photo description',
        position: PhotoPositionType.LEFT,
        color: null,
        media_id: 2
      }

      // act
      const promise = photoService.create(data)

      // assert
      expect(promise).rejects.toThrow('Can not create photo because media is already linked to a photo')
    })
  })

  describe('GET', () => {
    it('should get a photo by identifier', async () => {
      // act
      const photo = await photoService.getById(1)

      // assert
      expect(photo).toMatchSnapshot()
    })

    it('should throw a NotFoundError error when getting a photo with invalid id', async () => {
      // act
      const promise = photoService.getById(1000)

      // arrange
      await expect(promise).rejects.toThrow('Not found photo')
    })

    it('should get a photo by media identifier', async () => {
      // act
      const photo = await photoService.getByMediaId(1)

      // assert
      expect(photo).toMatchSnapshot()
    })

    it('should not get a photo by invalid media identifier', async () => {
      // act
      const photo = await photoService.getByMediaId(1000)

      // assert
      expect(photo).toBeNull()
    })

    it('should get the latest created photo', async () => {
      // act
      const photo = await photoService.getPreviousPhoto()

      // assert
      expect(photo).toMatchSnapshot()
    })

    it('should get paginated photos', async () => {
      // act
      const paginatedPhotos = await photoService.getPaginatedPhotos(1)

      //assert
      expect(paginatedPhotos).toMatchSnapshot()
    })

    it('should throw a NotFoundError error when getting paginated photos with invalid page', async () => {
      // act
      const promise = photoService.getPaginatedPhotos(1000)

      //assert
      expect(promise).rejects.toThrow('Page photo not found')
    })
  })

  describe('UPDATE', () => {
    it('should update photo successfully', async () => {
      // arrange
      const data = {
        title: 'photo title',
        description: 'photo description',
        color: '#000000',
        position: PhotoPositionType.RIGHT
      }

      // act
      const updatedPhoto = await photoService.update(1, data)

      // assert
      expect(updatedPhoto).toMatchSnapshot({
        updated_at: expect.any(Date)
      })
    })

    it('should throw a BadRequestError error when updating a photo with invalid media_id (media doesnt exist)', async () => {
      // arrange
      const data = {
        title: 'photo title',
        description: 'photo description',
        color: '#000000',
        position: PhotoPositionType.RIGHT,
        media_id: 1000
      }

      // act
      const promise = photoService.update(1, data)

      // assert
      expect(promise).rejects.toThrow('Can not update photo because media does not exist')
    })

    it('should throw a BadRequestError error when updating a photo with invalid media_id (media is already linked with a photo)', async () => {
      // arrange
      const data = {
        title: 'photo title',
        description: 'photo description',
        color: '#000000',
        position: PhotoPositionType.RIGHT,
        media_id: 2
      }

      // act
      const promise = photoService.update(1, data)

      // assert
      expect(promise).rejects.toThrow('Can not create photo because media is already linked to a photo')
    })
  })

  describe('DELETE', () => {
    it('should delete photo successfully', async () => {
      // arrange
      const db = { query: jest.fn() }

      // act
      await photoService.delete(1, db)

      // assert
      expect(photoService.getById(1)).rejects.toThrow('Not found photo')
    })

    it('should throw a NotFoundError error when deleting photo with invalid id', async () => {
      // arrange
      const db = { query: jest.fn() }

      // assert
      expect(photoService.delete(1000, db)).rejects.toThrow('Not found photo')
    })
  })
})
