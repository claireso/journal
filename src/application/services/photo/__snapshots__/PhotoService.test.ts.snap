// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application/PhotoService CREATE should create a photo successfully 1`] = `
{
  "color": null,
  "created_at": 2024-10-05T17:12:40.400Z,
  "description": "new photo description",
  "id": 4,
  "media": {
    "created_at": 2024-10-05T17:12:40.400Z,
    "id": 4,
    "name": "jkl.jpg",
    "size": {
      "height": 1093,
      "width": 1640,
    },
    "type": "image",
  },
  "media_id": 4,
  "name": "jkl.jpg",
  "portrait": false,
  "position": "left",
  "square": false,
  "title": "new photo",
  "updated_at": 2024-10-05T17:12:40.400Z,
}
`;

exports[`application/PhotoService GET should get a photo by identifier 1`] = `
{
  "color": null,
  "created_at": 2024-10-05T17:12:40.400Z,
  "description": "",
  "id": 1,
  "media": {
    "created_at": 2024-10-05T17:12:40.400Z,
    "id": 1,
    "name": "abc.jpg",
    "size": {
      "height": 1093,
      "width": 1640,
    },
    "type": "image",
  },
  "media_id": 1,
  "name": "abc.jpg",
  "portrait": false,
  "position": "left",
  "square": false,
  "title": "",
  "updated_at": 2024-10-05T17:12:40.400Z,
}
`;

exports[`application/PhotoService GET should get a photo by media identifier 1`] = `
{
  "color": null,
  "created_at": 2024-10-05T17:12:40.400Z,
  "description": "",
  "id": 1,
  "media": {
    "created_at": 2024-10-05T17:12:40.400Z,
    "id": 1,
    "name": "abc.jpg",
    "size": {
      "height": 1093,
      "width": 1640,
    },
    "type": "image",
  },
  "media_id": 1,
  "name": "abc.jpg",
  "portrait": false,
  "position": "left",
  "square": false,
  "title": "",
  "updated_at": 2024-10-05T17:12:40.400Z,
}
`;

exports[`application/PhotoService GET should get paginated photos 1`] = `
{
  "items": [
    {
      "color": null,
      "created_at": 2024-10-05T17:12:40.400Z,
      "description": "",
      "id": 1,
      "media": {
        "created_at": 2024-10-05T17:12:40.400Z,
        "id": 1,
        "name": "abc.jpg",
        "size": {
          "height": 1093,
          "width": 1640,
        },
        "type": "image",
      },
      "media_id": 1,
      "name": "abc.jpg",
      "portrait": false,
      "position": "left",
      "square": false,
      "title": "",
      "updated_at": 2024-10-05T17:12:40.400Z,
    },
    {
      "color": null,
      "created_at": 2024-10-05T17:12:40.400Z,
      "description": "",
      "id": 2,
      "media": {
        "created_at": 2024-10-05T17:12:40.400Z,
        "id": 2,
        "name": "def.jpg",
        "size": {
          "height": 1200,
          "width": 800,
        },
        "type": "image",
      },
      "media_id": 2,
      "name": "def.jpg",
      "portrait": true,
      "position": "center",
      "square": false,
      "title": "",
      "updated_at": 2024-10-05T17:12:40.400Z,
    },
    {
      "color": null,
      "created_at": 2024-10-05T17:12:40.400Z,
      "description": "",
      "id": 3,
      "media": {
        "created_at": 2024-10-05T17:12:40.400Z,
        "id": 3,
        "name": "ghi.jpg",
        "size": {
          "height": 800,
          "width": 800,
        },
        "type": "image",
      },
      "media_id": 3,
      "name": "ghi.jpg",
      "portrait": false,
      "position": "center",
      "square": true,
      "title": "",
      "updated_at": 2024-10-05T17:12:40.400Z,
    },
  ],
  "pager": {
    "count": 3,
    "limit": 10,
    "offset": 0,
    "totalPages": 1,
  },
}
`;

exports[`application/PhotoService GET should get the latest created photo 1`] = `
{
  "color": null,
  "created_at": 2024-10-05T17:12:40.400Z,
  "description": "",
  "id": 3,
  "media": {
    "created_at": 2024-10-05T17:12:40.400Z,
    "id": 3,
    "name": "ghi.jpg",
    "size": {
      "height": 800,
      "width": 800,
    },
    "type": "image",
  },
  "media_id": 3,
  "name": "ghi.jpg",
  "portrait": false,
  "position": "center",
  "square": true,
  "title": "",
  "updated_at": 2024-10-05T17:12:40.400Z,
}
`;

exports[`application/PhotoService UPDATE should update photo successfully 1`] = `
{
  "color": "#000000",
  "created_at": 2024-10-05T17:12:40.400Z,
  "description": "photo description",
  "id": 1,
  "media": {
    "created_at": 2024-10-05T17:12:40.400Z,
    "id": 1,
    "name": "abc.jpg",
    "size": {
      "height": 1093,
      "width": 1640,
    },
    "type": "image",
  },
  "media_id": 1,
  "name": "abc.jpg",
  "portrait": false,
  "position": "right",
  "square": false,
  "title": "photo title",
  "updated_at": Any<Date>,
}
`;
